U/danielsf/imsim psf

danielsf opened the pull request:
This pull requests implements the new PSF models that we discussed in the SSim telecons.

@rmjarvis, @RobertLuptonTheGood, @cwwalter, @rmandelb, @dkirkby, @jchiang87 

(Apparently, I am only able to officially assign the PR to LSST group members, so I can't assign @rmjarvis...)

dkirkby commented on python/lsst/sims/GalSimInterface/galSimPSF.py:
```diff
@@ -155,7 +156,49 @@ def __init__(self, fwhm=0.6):
+        """
+        # This code was provided by David Kirkby in a private communication
+
+        wlen_eff = dict(u=365.49, g=480.03, r=622.20, i=754.06, z=868.21, y=991.66)[band]
```
Mention that these eff. wlens are from Table 2 of LSE-40 and y=y2.

dkirkby commented on python/lsst/sims/GalSimInterface/galSimPSF.py:
```diff
@@ -155,7 +156,49 @@ def __init__(self, fwhm=0.6):
+        # This code was provided by David Kirkby in a private communication
+
+        wlen_eff = dict(u=365.49, g=480.03, r=622.20, i=754.06, z=868.21, y=991.66)[band]
+        FWHMatm = rawSeeing * (wlen_eff / 500.) ** -0.3 * airmass ** 0.6
```
From LSST-20160 eqn (4.1)

dkirkby commented on python/lsst/sims/GalSimInterface/galSimPSF.py:
```diff
@@ -155,7 +156,49 @@ def __init__(self, fwhm=0.6):
+
+        wlen_eff = dict(u=365.49, g=480.03, r=622.20, i=754.06, z=868.21, y=991.66)[band]
+        FWHMatm = rawSeeing * (wlen_eff / 500.) ** -0.3 * airmass ** 0.6
+        FWHMsys = 0.4 * airmass ** 0.6
```
From LSST-20160 eqn (4.2).  Could change 0.4 to sqrt(0.25 ** 2 + 0.30 ** 2 + 0.08 ** 2) for completeness.

rmjarvis commented on python/lsst/sims/GalSimInterface/galSimPSF.py:
```diff
@@ -155,7 +156,49 @@ def __init__(self, fwhm=0.6):
 
-        gaussian1 = galsim.Gaussian(sigma=alpha)
-        gaussian2 = galsim.Gaussian(sigma=2.0*alpha)
+        rms_pixel_size = pixel_scale*pixel_scale/6.0
```
Here is a python session with GalSIm confirming that convolving by a pixel adds 1/12 scale^2 to sigma^2.  Not 1/6.
```
>>> im1 = galsim.ImageD(256,256,scale=0.1)
>>> im2 = im1.copy()
>>> gaus = galsim.Gaussian(sigma=1.7)
>>> gaus_with_pixel = galsim.Convolve(gaus, galsim.Pixel(scale=1))
>>> im1 = gaus.drawImage(im1, method='no_pixel')
>>> im2 = gaus_with_pixel.drawImage(im2, method='no_pixel')
>>> sig1 = im1.FindAdaptiveMom().moments_sigma * 0.1
>>> sig2 = im2.FindAdaptiveMom().moments_sigma * 0.1
>>> sig1
1.7000000000000002
>>> sig2
1.724542808532715
>>> math.sqrt(sig2**2 - 1./12)
1.7002101532247627
>>> 1/(sig2**2 - sig1**2)
11.897977442627738
```

danielsf replied:
Okay.  Thanks for the explanation.

cwwalter replied:
The 1/6th came from @RobertLuptonTheGood during our meeting.  I was also wondering about the sqrt(2) factor.  Let's have him confirm just to make sure.

rmjarvis replied:
I suspect he was thinking that T = Ixx + Iyy = scale^2/6 for a pixel.  I'm pretty sure he just forgot that the same quantity for a Gaussian is 2sigma^2, not just sigma^2.

RobertLuptonTheGood replied:
More or less.  I agree that my factor 1/6 was incorrect for the reasons given.

danielsf commented on tests/testFWHM.py:
```diff
@@ -86,71 +84,59 @@ class characterizing the detector corresponding to this image
-                  (distanceToLeft+distanceToRight, fwhm, maxFlux, theta/np.pi)
-
-            self.assertLess(np.abs(distanceToLeft+distanceToRight-fwhm), 0.1*fwhm, msg=msg)
+        self.assertLess(np.abs(fwhm-fwhm_best), 0.1*fwhm)
```
Are we sure that, given these images are driven with photon shooting, the test will ever be able to detect the difference between PSFs with and without pixel convolution?  We can either test in the well-sampled regime, in which case 1/12 of a pixel is inconsequential, or we can test in the poorly-sampled regime, in which case the stochasticity of photon shooting will overwhelm 1/12 of the pixel.  That, at least, was my experience when writing the test as-is.  The test was good enough to tell that the PSF was "about right", but it didn't really matter whether the 1/12 was subtracted, or not.  I thought that was due to the nature of photon shooting.  It could mean I was doing something wrong.

rmjarvis replied:
Just shoot more photons.  Photon shooting is as noise-free as you need it to be.  Just takes longer, but I think you should be able to get to pretty low noise without being unreasonably time consuming.

danielsf commented:
I believe I have addressed all of the concerns.

The problem with the FWHM unit test wasn't insufficient photons.  The test was not learning the pixel scale correctly.  I now just pass the true pixel scale in by hand.

I also added a test to validate the FWHM of the double Gaussian PSF without photon shooting.
